glusterd2/commands/volumes/volume-edit.go

```diff
@@ -0,0 +1,69 @@
+package volumecommands
+
+import (
+	"net/http"
+	"strings"
+
+	"github.com/gluster/glusterd2/glusterd2/gdctx"
+	restutils "github.com/gluster/glusterd2/glusterd2/servers/rest/utils"
+	"github.com/gluster/glusterd2/glusterd2/transaction"
+	"github.com/gluster/glusterd2/glusterd2/volume"
+	"github.com/gluster/glusterd2/pkg/api"
+	"github.com/gluster/glusterd2/pkg/errors"
+
+	"github.com/gorilla/mux"
+)
+
+func volumeEditHandler(w http.ResponseWriter, r *http.Request) {
+
+	p := mux.Vars(r)
+	volname := p["volname"]
+
+	ctx := r.Context()
+	logger := gdctx.GetReqLogger(ctx)
+
+	var req api.VolEditReq
+	if err := restutils.UnmarshalRequest(r, &req); err != nil {
+		restutils.SendHTTPError(ctx, w, http.StatusBadRequest, err)
+		return
+	}
+
+	//Lock on Volume Name
+	lock, unlock := transaction.CreateLockFuncs(volname)
+	if err := lock(ctx); err != nil {
+		if err == transaction.ErrLockTimeout {
+			restutils.SendHTTPError(ctx, w, http.StatusConflict, err)
+		} else {
+			restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
+		}
+		return
+	}
+	defer unlock(ctx)
+
+	//validate volume name
+	v, err := volume.GetVolume(volname)
+	if err != nil {
+		if err == errors.ErrVolNotFound {
+			restutils.SendHTTPError(ctx, w, http.StatusNotFound, errors.ErrVolNotFound)
+		} else {
+			restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
+		}
+		return
+	}
+
+	for key, value := range req.Metadata {
+		if strings.HasPrefix(key, "_") {
+			logger.WithField("key", key).Error("Key names starting with '_' are restricted in metadata field")
+			restutils.SendHTTPError(ctx, w, http.StatusBadRequest, "Key names starting with '_' are restricted in metadata field")
+			return
+		}
+		v.Metadata[key] = value
+	}
+	if err := volume.AddOrUpdateVolumeFunc(v); err != nil {
+		logger.WithError(err).WithField(
+			"volume", v.Name).Debug("failed to store volume info")
```

Comment:
This is a serious issue - as there is no response sent back to the client, it will hang.

Reply:
Added a http error response.